### PR TITLE
Cmake support static builds

### PR DESCRIPTION
**Pull request by drwells**
As usual the checklist doesn't apply.

This change also changes the link order to better handle our dependencies that are statically linked and depend on HDF5.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -745,6 +748,21 @@ ENDMACRO()
 # everything but the source files)
 FUNCTION(IBAMR_SETUP_TARGET_LIBRARY target_library)
   MESSAGE(STATUS "setting up target ${target_library}")
+
+  # Figure out if we are a 2D or a 3D library:
```
Comment (boyceg) [inline CMakeLists.txt:752]: Can one configuration build both 2D and 3D? Or does a particular configuration now have its spatial dimension fixed at configure time?

File: doc/cmake.md [hunk 2]
```diff
@@ -79,6 +83,16 @@ problems with CMake, a good way to get out of trouble is to delete
 `CMakeCache.txt`, which will delete the cache. Deleting the cache is harmless -
 the subsequent initial configuration run will just take a few more seconds.
 
+### Configuring the build
+- If you want to build IBAMR with static libraries then pass the argument
+  `-DBUILD_SHARED_LIBS=ON` to the initial call to `cmake`. IBAMR defaults to
```
Comment (boyceg) [inline doc/cmake.md:88]: ON or OFF?

Comment (drwells) [reply doc/cmake.md:88]: oops :blush:

File: CMakeLists.txt (hunk 1 above)
Comment (drwells) [reply CMakeLists.txt:752]: The default is to build both but we still compile IBAMR2d, IBAMR3d, IBTK2d, IBTK3d as independent libraries.

Comment (drwells) [reply CMakeLists.txt:752]: This particular bit of logic refers to if we should link to the 2D or 3D versions of SAMRAI.

**Review by boyceg — APPROVED** [review]

**Pull request merged**